src/libraries/System.Private.CoreLib/src/System/Threading/WaitHandle.Windows.cs

```diff
@@ -56,40 +56,56 @@ private static unsafe int WaitForMultipleObjectsIgnoringSyncContext(IntPtr* pHan
+#endif
+
+                while (true)
                 {
-                    result = (int)Interop.Kernel32.WaitForMultipleObjectsEx((uint)numHandles, (IntPtr)pHandles, waitAll ? Interop.BOOL.TRUE : Interop.BOOL.FALSE, (uint)millisecondsTimeout, Interop.BOOL.TRUE);
-                }
+#if NATIVEAOT
+                    if (reentrantWait)
+                    {
+                        Debug.Assert(!waitAll);
+                        result = RuntimeImports.RhCompatibleReentrantWaitAny(true, millisecondsTimeout, numHandles, pHandles);
+                    }
+                    else
+                    {
+                        result = (int)Interop.Kernel32.WaitForMultipleObjectsEx((uint)numHandles, (IntPtr)pHandles, waitAll ? Interop.BOOL.TRUE : Interop.BOOL.FALSE, (uint)millisecondsTimeout, Interop.BOOL.TRUE);
+                    }
 #else
-                result = (int)Interop.Kernel32.WaitForMultipleObjectsEx((uint)numHandles, (IntPtr)pHandles, waitAll ? Interop.BOOL.TRUE : Interop.BOOL.FALSE, (uint)millisecondsTimeout, Interop.BOOL.TRUE);
+                    result = (int)Interop.Kernel32.WaitForMultipleObjectsEx((uint)numHandles, (IntPtr)pHandles, waitAll ? Interop.BOOL.TRUE : Interop.BOOL.FALSE, (uint)millisecondsTimeout, Interop.BOOL.TRUE);
 #endif
 
-                if (result != Interop.Kernel32.WAIT_IO_COMPLETION)
-                    break;
+                    if (result != Interop.Kernel32.WAIT_IO_COMPLETION)
+                        break;
 
-                // Handle APC completion by adjusting timeout and retrying
-                if (millisecondsTimeout != -1)
-                {
-                    long currentTime = Environment.TickCount64;
-                    long elapsed = currentTime - startTime;
-                    if (elapsed >= millisecondsTimeout)
+#if !MONO
+                    // Check for thread interrupt when APC completion occurs
+                    Thread.CheckForPendingInterrupt();
+#endif
+
+                    // Handle APC completion by adjusting timeout and retrying
+                    if (millisecondsTimeout != -1)
                     {
-                        result = Interop.Kernel32.WAIT_TIMEOUT;
-                        break;
+                        long currentTime = Environment.TickCount64;
+                        long elapsed = currentTime - startTime;
+                        if (elapsed >= millisecondsTimeout)
+                        {
+                            result = Interop.Kernel32.WAIT_TIMEOUT;
+                            break;
+                        }
+                        millisecondsTimeout -= (int)elapsed;
+                        startTime = currentTime;
                     }
-                    millisecondsTimeout -= (int)elapsed;
-                    startTime = currentTime;
                 }
             }
-            currentThread.ClearWaitSleepJoinState();
+            finally
+            {
+                currentThread.ClearWaitSleepJoinState();
```

Comment:
If somebody creates an exception filter and calls Wait inside the filter, the exception filter is going to be called before this finally and `Debug.Assert(GetThreadStateBit(ThreadState.WaitSleepJoin));` inside ClearWaitSleepJoinState is going to fail.

It may be better to clear ClearWaitSleepJoinState inside CheckForPendingInterrupt if it is set, and keep this as non-exceptional path.